api/src/auth/drivers/saml.ts

```diff
@@ -0,0 +1,178 @@
+samlify.setSchemaValidator(validator);
+
+export class SAMLAuthDriver extends LocalAuthDriver {
+	idp: any;
+	sp: any;
+	redirectUrl: string;
+	usersService: UsersService;
+	config: Record<string, any>;
+
+	constructor(options: AuthDriverOptions, config: Record<string, any>) {
+		super(options, config);
+
+		this.config = config;
+		this.usersService = new UsersService({ knex: this.knex, schema: this.schema });
+
+		const redirectUrl = new Url(env.PUBLIC_URL).addPath('auth', 'login', config.provider, 'callback');
+		this.redirectUrl = redirectUrl.toString();
+
+		this.sp = samlify.ServiceProvider(getConfigFromEnv(`AUTH_${config.provider.toUpperCase()}_SP`));
+		this.idp = samlify.IdentityProvider(getConfigFromEnv(`AUTH_${config.provider.toUpperCase()}_IDP`));
+	}
+
+	async fetchUserID(identifier: string) {
+		const user = await this.knex
+			.select('id')
+			.from('directus_users')
+			.whereRaw('LOWER(??) = ?', ['external_identifier', identifier.toLowerCase()])
+			.first();
+
+		return user?.id;
+	}
+
+	async getUserID(payload: Record<string, any>) {
+		const { provider, emailKey, identifierKey, nameKey, givenNameKey, familyNameKey, allowPublicRegistration } =
+			this.config;
+
+		const email = payload[emailKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress'];
+		const identifier = payload[identifierKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier'];
+		const userID = await this.fetchUserID(identifier);
+
+		const userInfo = {
+			name: payload[nameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name'],
+			given_name: payload[givenNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname'],
+			family_name: payload[familyNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname'],
+			email,
+		};
+
+		if (userID) return userID;
+
+		// Is public registration allowed?
+		if (!allowPublicRegistration) {
+			logger.trace(`[SAML] User doesn't exist, and public registration not allowed for provider "${provider}"`);
+			throw new InvalidCredentialsException();
+		}
+
+		try {
+			await this.usersService.createOne({
+				provider,
+				first_name: userInfo.given_name,
+				last_name: userInfo.family_name,
+				email: email,
+				external_identifier: identifier.toLowerCase(),
+				role: this.config.defaultRoleId,
+			});
+			const userID = await this.fetchUserID(identifier);
+			if (userID) return userID;
+		} catch (e) {
+			if (e instanceof RecordNotUniqueException) {
+				logger.warn(e, '[SAML] Failed to register user. User not unique');
+				throw new InvalidProviderException();
+			}
+			throw e;
+		}
+
+		throw new InvalidCredentialsException();
+	}
+
+	async login(user: User): Promise<void> {
+		let authData = user.auth_data as AuthData;
+
+		if (typeof authData === 'string') {
+			try {
+				authData = parseJSON(authData);
+			} catch {
+				logger.warn(`[SAML] Session data isn't valid JSON: ${authData}`);
+			}
+		}
+	}
```

Comment:
I don't quite understand this function. It seems we never set `auth_data`. Is this ever used, and if so, wouldn't it be better to throw is the data was invalid?

Reply:
We don't use it you can correct, can change to just a throw instead.

Reply:
But when do we set `auth_data`? This should never trigger, right?